src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocketOptions.cs

```diff
@@ -0,0 +1,167 @@
+
+        // Note that some headers are restricted like Host.
+        public void SetRequestHeader(string headerName, string headerValue)
+        {
+            ThrowIfReadOnly();
+
+            // WebHeaderCollection performs validation of headerName/headerValue.
+            _requestHeaders[headerName] = headerValue;
+        }
+
+        internal WebHeaderCollection RequestHeaders { get { return _requestHeaders; } }
+
+        public ICredentials Credentials
+        {
+            get
+            {
+                return _credentials;
+            }
+            set
+            {
+                ThrowIfReadOnly();
+                _credentials = value;
+            }
+        }
+
+        public IWebProxy Proxy
+        {
+            get
+            {
+                return _proxy;
+            }
+            set
+            {
+                ThrowIfReadOnly();
+                _proxy = value;
+            }
+        }
+
+        [SuppressMessage("Microsoft.Usage", "CA2227:CollectionPropertiesShouldBeReadOnly",
+            Justification = "This collection will be handed off directly to HttpWebRequest.")]
+        public X509CertificateCollection ClientCertificates
+        {
+            get
+            {
+                if (_clientCertificates == null)
+                {
+                    _clientCertificates = new X509CertificateCollection();
+                }
+                return _clientCertificates;
+            }
+            set
+            {
+                ThrowIfReadOnly();
+                if (value == null)
+                {
+                    throw new ArgumentNullException("value");
+                }
+                _clientCertificates = value;
+            }
+        }
+
+        public CookieContainer Cookies
+        {
+            get
+            {
+                return _cookies;
+            }
+            set
+            {
+                ThrowIfReadOnly();
+                _cookies = value;
+            }
+        }
+
+        #endregion HTTP Settings
+
+        #region WebSocket Settings
+
+        public void AddSubProtocol(string subProtocol)
+        {
+            ThrowIfReadOnly();
+            WebSocketValidate.ValidateSubprotocol(subProtocol);
+
+            // Duplicates not allowed.
+            foreach (string item in _requestedSubProtocols)
+            {
+                if (string.Equals(item, subProtocol, StringComparison.OrdinalIgnoreCase))
+                {
+                    throw new ArgumentException(SR.Format(SR.net_WebSockets_NoDuplicateProtocol, subProtocol),
+                        "subProtocol");
+                }
+            }
+            _requestedSubProtocols.Add(subProtocol);
+        }
+
+        public TimeSpan KeepAliveInterval
+        {
+            get
+            {
+                return _keepAliveInterval;
+            }
+            set
+            {
+                ThrowIfReadOnly();
+                if (value < Timeout.InfiniteTimeSpan)
```

Comment:
I understand what this is doing, but FWIW I find it confusing.  When I see "Infinite", I immediatley think this must be a max value, such that throwing for anything less than that would be strange.  And then I remind myself it's -1 :smile: